src/modules/ctc/src/ctc.cc

```diff
@@ -17,6 +17,191 @@
+    {
+        train->authority.pop();
+        types::BlockId current_block_id = train->authority.front();
+        types::Block   current_block    = GetBlockById(current_block_id);
+        train->suggested_speed = current_block.speed_limit;
+    }
+    else
+    {
+        error = types::ERROR_INVALID_TRAIN;
+    }
+    return error;
+}
+
+/*------------------------------------- Setters -------------------------------------*/
+void Ctc::SetBlocks(std::vector<types::Block> &blocks)
+{
+    types::Block yard;
+    yard.block = 0;
+    blocks_.push_back(yard);
+    for ( types::Block &block : blocks)
+    {
+        blocks_.push_back(block);
+    }
+}
+
+void Ctc::SetStations(std::vector<types::Block> &blocks)
+{
+    for (types::Block block : blocks)
+    {
+        if (block.has_station)
+        {
+            ctc::Station station(block.station_name, block.block, block.total_time_to_station);
+            stations_.push_back(station);
+        }
+    }
+}
+
+void Ctc::SetScheduleFilePath(std::filesystem::path path)
+{
+    schedule_file_path_ = path;
+}
+
+void Ctc::SetDefaultRoute(void)
+{
+    types::BlockId yard = 0;
+    default_route_.push_back(yard);
+    // Add K, L, M, N, O, P, Q
+    for (std::size_t i = 63; i < 101; ++i)
+    {
+        default_route_.push_back(blocks_[i].block);
+    }
+    // Add N
+    for (std::size_t i = 85; i > 76; --i)
+    {
+        default_route_.push_back(blocks_[i].block);
+    }
+    for (std::size_t i = 101; i < blocks_.size(); ++i)
+    {
+        default_route_.push_back(blocks_[i].block);
+    }
+    for (std::size_t i = 28; i > 0; --i)
+    {
+        default_route_.push_back(blocks_[i].block);
+    }
+    for (std::size_t i = 13; i < 58; ++i)
+    {
+        default_route_.push_back(blocks_[i].block);
+    }
+    default_route_.push_back(yard);
+}
+
+void Ctc::SetManualMode(void)
+{
+    ctc_mode_ = CtcOperationMode::MANUAL_MODE;
+}
+
+/*------------------------------------- Getters -------------------------------------*/
+types::Block Ctc::GetBlockById(const types::BlockId block_id) const
+{
+    types::Block result;
+    for (types::Block block : blocks_)
+    {
+        if (block.block == block_id)
+        {
+            result = block;
+        }
+    }
+    return result;
+}
+
+std::size_t Ctc::GetNumStation(void) const
+{
+    return stations_.size();
+}
+
+std::vector<ctc::Station> Ctc::GetStations(void) const
+{
+    return stations_;
+}
+
+std::vector<types::BlockId> Ctc::GetDefaultRoute(void) const
+{
+    return default_route_;
+}
+
+std::vector<types::BlockId> Ctc::GetRoute(const types::BlockId destination)
+{
+    std::vector<types::BlockId> route;
+    for (size_t i = 0; i < default_route_.size(); ++i)
+    {
+        if (default_route_[i] == destination)
+        {
+            std::copy(default_route_.begin(), default_route_.begin() + i + 1, std::back_inserter(route));
+        }
+    }
+    return route;
+}
+
+ctc::Train* Ctc::GetTrainPointerById(const types::TrainId train_id)
```

Comment:
Again, make sure any functions that are only called internally are private.  Especially ones that potentially allow private data inside the class to be mutated outside the class.  Consider using a reference or smart pointer for any functions like this that return a raw pointer.  Another approach is using `find_if` with a lambda function.